flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/JoinedRowData.java

```diff
@@ -16,47 +16,73 @@
  * limitations under the License.
  */
 
-package org.apache.flink.table.data.utils;
-
-import org.apache.flink.annotation.Internal;
-import org.apache.flink.table.data.ArrayData;
-import org.apache.flink.table.data.DecimalData;
-import org.apache.flink.table.data.MapData;
-import org.apache.flink.table.data.RawValueData;
-import org.apache.flink.table.data.RowData;
-import org.apache.flink.table.data.StringData;
-import org.apache.flink.table.data.TimestampData;
+package org.apache.flink.table.data;
+
+import org.apache.flink.annotation.PublicEvolving;
 import org.apache.flink.types.RowKind;
 
+import javax.annotation.Nullable;
+
 import java.util.Objects;
 
-/** An implementation of {@link RowData} which is backed by two concatenated {@link RowData}. */
-@Internal
+/**
+ * An implementation of {@link RowData} which is backed by two concatenated {@link RowData}.
+ *
+ * <p>This implementation is mutable to allow for performant changes in hot code paths.
+ */
+@PublicEvolving
 public class JoinedRowData implements RowData {
 
     private RowKind rowKind = RowKind.INSERT;
     private RowData row1;
```

Comment:
(I didn't add `@Nullable` to the fields since this produces a lot of warnings in the class, and making the IDE happy by checking non-nullability first impose a runtime cost. I decided to instead document this explicitly.)